packages/host/app/services/loader-service.ts

```diff
@@ -98,11 +98,29 @@ export default class LoaderService extends Service {
     } else {
       body = await request.text();
     }
-    return await this.loader.fetch(request.url, {
+    let response = await this.loader.fetch(request.url, {
       method: request.method,
       headers: new Headers(request.headers),
       body,
     });
+
+    // Obtaining a new JWT
+    // if receiving `Authorization` error,
+    // which means user permissions is updated from the server.
+    if (!response.ok && response.status === 403) {
```

Comment:
From these lines:
https://github.com/cardstack/boxel/blob/21f07d75137958b1f0312fd148b7ab10e2f80a11/packages/runtime-common/realm.ts#L995

https://github.com/cardstack/boxel/blob/21f07d75137958b1f0312fd148b7ab10e2f80a11/packages/runtime-common/realm.ts#L814

It seems we only need to retrieve a new JWT token if the response status is 403 because the error related to user permission is `Authorization` error. Please correct me if I'm wrong.

Reply:
403 is when the authenticated user has no permission to read the realm. I'm having difficulties understanding how refreshing the token would help with this, am I missing something?

Reply:
It may not help. But we would should be returning this if the token has expired, or if the user has been granted _more_ rights than the token they provide has: eg the user has a token for read only but write perms were added to the realm, and now the user is trying to write to the realm. When there is a mismatch between the token and the realm permissions a 404 should be returned, it’s the servers way of saying “get a new token”

Reply:
@habdelra thanks for explaining the scenario, @jurgenwerk I added this checking to compare the permissions in current token and the permissions in the server, so we are not only returning 403 if user has insufficient permission but also if the permissions don't match. https://github.com/cardstack/boxel/pull/1057/files#diff-b6059baf94cff52ea5df0ab5f23724bc3c52a3685ea3a844fc1d9c9c84192bddR996

Reply:
I see. I am wondering if we could make a better separation for these cases:

1. Not allowed to perform the action
2. Token has expired
3. Token has invalid data 

If I understand correctly, all those are covered by returning a 403. But looking at this, only cases (2) and (3) are an actual way of telling the client that it has to refresh the token.

Reply:
@jurgenwerk Do you think it is better to use different status code for the error that requires client to refresh the token?

Reply:
Based on discussion with @jurgenwerk , we decided to use 401, so if we receive 403 we don't need to retrieve a new token. But any other advices are welcomed.